internal/core/worker/operationexecution/operation_execution_worker.go

```diff
@@ -161,65 +167,77 @@ func (w *OperationExecutionWorker) executeOperationFlow(operationID string) erro
-		w.rollbackOperation(op, def, executionErr, loopLogger, executor)
+		w.handleExecutionError(op, def, executionErr, loopLogger, executor)
 	} else {
 		op.Status = operation.StatusFinished
+		w.finishOperationAndLease(w.workerContext, op, def, loopLogger)
 	}
 
-	w.finishOperationAndLease(op, def, loopLogger)
-
 	return nil
 }
 
-func (w *OperationExecutionWorker) rollbackOperation(op *operation.Operation, def operations.Definition, executionErr error, loopLogger *zap.Logger, executor operations.Executor) {
-	w.operationManager.AppendOperationEventToExecutionHistory(w.workerContext, op, "Starting operation rollback")
+func (w *OperationExecutionWorker) rollbackOperation(ctx context.Context, op *operation.Operation, def operations.Definition, executionErr error, loopLogger *zap.Logger, executor operations.Executor) {
+	w.operationManager.AppendOperationEventToExecutionHistory(ctx, op, "Starting operation rollback")
 	rollbackErr := w.executeRollbackCollectingLatencyMetrics(op.DefinitionName, func() error {
-		return executor.Rollback(w.workerContext, op, def, executionErr)
+		return executor.Rollback(ctx, op, def, executionErr)
 	})
 
 	if rollbackErr != nil {
 		loopLogger.Error("operation rollback failed", zap.Error(rollbackErr))
-		w.operationManager.AppendOperationEventToExecutionHistory(w.workerContext, op, fmt.Sprintf("Operation rollback flow execution failed, reason: %s", rollbackErr.Error()))
+		w.operationManager.AppendOperationEventToExecutionHistory(ctx, op, fmt.Sprintf("Operation rollback flow execution failed, reason: %s", rollbackErr.Error()))
 	} else {
-		w.operationManager.AppendOperationEventToExecutionHistory(w.workerContext, op, "Operation rollback flow execution finished with success")
+		w.operationManager.AppendOperationEventToExecutionHistory(ctx, op, "Operation rollback flow execution finished with success")
 	}
 }
 
-func (w *OperationExecutionWorker) handleExecutionError(op *operation.Operation, executionErr error, loopLogger *zap.Logger) {
+func (w *OperationExecutionWorker) handleExecutionError(op *operation.Operation, def operations.Definition, executionErr error, loopLogger *zap.Logger, executor operations.Executor) {
+	executionErrCtx := w.workerContext
+	var msg string
 	if strings.Contains(executionErr.Error(), context.Canceled.Error()) {
 		op.Status = operation.StatusCanceled
-
-		loopLogger.Info("operation execution canceled")
-		w.operationManager.AppendOperationEventToExecutionHistory(w.workerContext, op, "Operation canceled by the user")
+		msg = "operation canceled by the user"
+		var cancelFunc func()
+		executionErrCtx, cancelFunc = context.WithTimeout(context.Background(), TerminationGracePeriodTimeout)
+		defer cancelFunc()
 	} else {
 		op.Status = operation.StatusError
-
-		loopLogger.Error("operation execution failed", zap.Error(executionErr))
-		w.operationManager.AppendOperationEventToExecutionHistory(w.workerContext, op, "Operation execution failed")
+		msg = "operation execution failed"
```

Comment:
when will this branch execute? an execution error that is not `executionErr.Error()` or `context.Canceled.Error()`

so what is it?